File: perl/improveMSA.pl
use FileUtil;
use Getopt::Long qw(:config no_ignore_case no_auto_abbrev);
use Pod::Usage;

# Copyright (c) 2021, Indiana University, Bloomington. All rights reserved.

=head1 NAME

improveMSA.pl - Takes a set of genus specific reads, cluster them at a specific percent
  identity cutoff, and generate an MSA and consensus sequence for each cluster.

=head1 VERSION

This document refers to version 1.00 of improveMSA.pl, released 12.27.2021.

=head1 SYNOPSIS

improveMSA.pl [-help]

Options:
    
    -log                  Print logging info to standard error
    -help|h               Print help message
    -reset                Overwrite existing results

    -msa                  Input MSA file
    -out                  Write output file here
    -rocitsPath           Full path to where RoC-ITS scripts are located

    -minimumCov           Minimum coverage required to keep a column in the MSA (default 0)
    -wordSize             Anchor size for conserved regions (default 12)
    -maxJobs              Max number of concurrent jobs (default 40)
    -local                Directory where temporary files will be written (default ".")

    -probcons             (Optional) Full path to probcons executable

=head1 DESCRIPTION

=head2 Overview

improveMSA.pl is designed to correct problems with muscle MSA. The probcons MSA tool tends to
correct these problems but is MUCH slower. Here we identify anchor regions (regions that are
nearly identical in the vast majority of reads) and then re-align the intervening bases. Anchors
must have a reasonable minimum lenght to be effective (8-20 bps). The larger the word size, the
fewer blocks of sequence there will be, and they will be appropriately larger slowing down the
re-alignment process. Too small and the blocks are not guaranteed to be meaningful and there is
a possibility that the alignment will be sub-optimal.

=cut

my ($help, $log, $input, $output, $word, $minCov, $jobs, $localDir);
my ($probconsBin, $rocitsPath);

BEGIN: {
  GetOptions(
    "help|h"        => \$help,
    "log"           => \$log,
    "msa=s"         => \$input,
    "out=s"         => \$output,
    "wordSize:s"    => \$word,
    "local:s"       => \$localDir,
    "rocitsPath=s"  => \$rocitsPath,
    "maxJobs:s"     => \$jobs,
    "minimumCov:s"  => \$minCov,
    "probcons:s"    => \$probconsBin,
  ) || pod2usage(2);
  pod2usage(1) if defined($help);
}
if (!defined($input)) {
  pod2usage(1);
  exit;
}

$word = 12 unless defined($word);
$minCov = 0 unless defined($minCov);
$localDir = "." unless defined($localDir);
$probconsBin = "probcons" unless defined($probconsBin);
$jobs = 40 unless defined($jobs);
my $gggs = "n" x 22;

my $mfh = FileUtil::openFileHandle("perl $rocitsPath/reformatFasta.pl $input 0 99999999|");
while (<$mfh>) {
  if (/^>(\S+)/) {
    my $id = $1;
    $_ = <$mfh>;
    chomp; 
    $seq{$id} = $gggs.uc($_).$gggs;
    my @seq = split //,$seq{$id};
    $len = @seq;
    push @seqs1,\@seq;
    push @ids,$id;
  }
}
$mfh->close();

my @good;
my $cnt = @ids;
for (my $i = 0; $i < $len; $i++) {
  my $bases = 0;
  for (my $j = 0; $j < $cnt; $j++) {
    $dist{${$seqs1[$j]}[$i]}++;
    if (${$seqs1[$j]}[$i] eq "-") {
    } else {
      $bases++;
    }
  }
  if ($bases < $minCov) {
    for (my $j = 0; $j < $cnt; $j++) {
      ${$seqs1[$j]}[$i] = "";
    }
  }
}
for (my $j = 0; $j < $cnt; $j++) {
  my $tmp = join "",@{$seqs1[$j]};
  my @seq = split //,$tmp;
  $len = @seq;
  push @seqs,\@seq;
}

for (my $i = 0; $i < $len; $i++) {
  my %dist;
  for (my $j = 0; $j < $cnt; $j++) {
    $dist{${$seqs[$j]}[$i]}++;
  }
  if ($dist{"-"}/$cnt >= 0.1) {
    push @good,"-";
  } else {
    my $best;
    foreach my $j (sort {$dist{$b}<=>$dist{$a}; } keys %dist) {
      $best = $j;
      last;
    }
    if ($dist{$best}/$cnt < 0.9) {
      push @good,"-";
    } else {
      push @good,"G";
    }
  }
}

my $g = join "",@good;
if ($g =~ /^(-+)/) {
  my $w = "p" x length($1);
  $g =~ s/^(-+)/$w/;
}

my @gg = split //,$g;
my @regionBegins;
my @regionEnds;
my $start = 0;
my $endFound = 0;
while (!$endFound) {
  my $i = $start;
  for ($i = $start; $i < @gg; $i++) {
    if ($gg[$i] eq "-") {
      $begin = $i - $word;
      my $tot = 0;
      my $pos = $i;
      while ($tot < $word) {
        if ($gg[$pos] eq "G") {
          $tot++;
        } else {
          $tot = 0;
        }
        $pos++;
      }
      $end = $pos;
#      my $x = join "",@gg[$begin..$end];
#      print "$i $begin $end\n$x\n";
      push @regionBegins,$begin;
      push @regionEnds,$end;
      $start = $end - $word;
      last;
    }
  }
  if ($i < @gg) {
  } else {
    $endFound = 1;
  }
}

for (my $ii = 0; $ii < @regionBegins; $ii++) {
  my $fho = new FileHandle(">$localDir/temp.$$.$ii.$regionBegins[$ii].$regionEnds[$ii].fa");
  for (my $i = 0; $i < $cnt; $i++) {
    my @set;
    for (my $j = $regionBegins[$ii]; $j < $regionEnds[$ii]; $j++) {
      my $chr = ${$seqs[$i]}[$j];
      if ($j < ($regionEnds[$ii]-$word)) {
        ${$seqs[$i]}[$j] = "";
        push @set,$chr;
      } else {
        ### subsequences that have gaps in the right anchor region are by definition rare but need to
        ### be replaced by N's. This is important as later non-dashes in the right anchor are removed
        ### to when the new msa is inserted.
        if ($chr eq "-") {
          push @set,"n";
        } else {
          push @set,$chr;
        }
      }
    }
    my $sub = join "",@set;
    $sub =~ s/-+//g;
    print $fho ">$ids[$i] $regionBegins[$ii] $regionEnds[$ii]\n";
    print $fho $sub,"\n";
  }
  $fho->close();
  my $probCmd = "$probconsBin -c 0 $localDir/temp.$$.$ii.$regionBegins[$ii].$regionEnds[$ii].fa |perl $rocitsPath/reformatFasta.pl - 0 9999999 > $localDir/temp.$$.$ii.$regionBegins[$ii].out &";
  push @cmds,$probCmd;
  print STDERR $probCmd,"\n" if $log;
}
my $ccnt = 0;
my $fho = new FileHandle(">$localDir/temp.$$.cmd");
foreach my $i (@cmds) {
  print $fho $i,"\n";
  $ccnt++;
  if ($cnt == $jobs) {
    print $fho "wait\n";
    $ccnt = 0;
  }
}
print $fho "wait\n";
$fho->close();

{
  ### have to go to some lengths to redirect stderr not to clutter output
  my $cmd = "tcsh -x $localDir/temp.$$.cmd > /dev/null";
  print STDERR $cmd,"\n";
  open(my $STDOLD, '>&', STDERR);
  open(STDERR, '>>', "/dev/null");
  my $run = system($cmd);
  open(STDERR, '>&', $STDOLD);
}

for (my $ii = 0; $ii < @regionBegins; $ii++) {
  my @result;
  my $rfh = FileUtil::openFileHandle("$localDir/temp.$$.$ii.$regionBegins[$ii].out");
  while (<$rfh>) {
    next if /^>/;
    chomp;
    push @result,$_;
  }
  $rfh->close();
  for (my $i = 0; $i < $cnt; $i++) {
    for (my $j = $regionBegins[$ii]; $j < $regionEnds[$ii]; $j++) {
      my @res = split //,$result[$i];
      my $e = @res;
      my $removed = 0;
      for (my $k = $e - 1; $k >= 0; $k--) {
        if ($res[$k] =~ /\w/o) {
          $res[$k] = "";
          $removed++;
        }
        last if $removed == $word;
      }
      my $res = join "",@res;
      ${$seqs[$i]}[$j] = lc($res);
      last;
    }
  }
}

my $fho = new FileHandle(">$output");
for (my $i = 0; $i < $cnt; $i++) {
  my $l = join "",@{$seqs[$i]};
  $l =~ s/n/-/g;
  print $fho ">$ids[$i]\n$l\n";
}
$fho->close();

$cmd = "rm $localDir/temp.$$*.fa";
system($cmd);
$cmd = "rm $localDir/temp.$$*.out $localDir/temp.$$.cmd";
system($cmd);

exit;

File: python/improve-msa.py
from collections import Counter
import os
import multiprocessing as mp
from Bio import SeqIO, SeqRecord, Seq
from Bio.SeqIO import FastaIO
import argparse
import sys
import tempfile
import subprocess
import re
import time

if sys.version_info[0] < 3:
    raise Exception("Python 3 or a more recent version is required.")

parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument('--msa', help='MSA containing processed subreads')
parser.add_argument('--max-cores', type=int, default=4, help='Maximum number of cores to use')
parser.add_argument('--scan', action='store_true', help='Print number of regions and exit')
parser.add_argument('--benchmark', action='store_true',
                    help='Print the time script starts and ends to the project log file')
parser.add_argument('--pipeline', action='store_true',
                    help='Pipeline base path [project/prefix/subread]')
parser.add_argument('--base', help='Output base. Required for pipeline or benchmark')
parser.add_argument('--word-size', type=int, default=12,
                    help='Word size to use')
args = parser.parse_args()

# configure based on if part of a pipeline
if args.pipeline:
    if args.base is None:
        parser.error('You must provide --base with --pipeline')
    args.max_cores = 1
    args.benchmark = True
    args.msa = f'{args.base}-corrected-smushed.best.fas'
    if os.path.exists(f'{args.base}-improve-msa.benchmark'):
        quit()
    output = open(f'{args.base}-smushed-improved.fa', 'w')

else:
    if args.msa is None:
        parser.error('You must provide either --msa or --pipeline')
    if args.benchmark and args.base is None:
        parser.error('You must provide --base with --benchmark')
    output = sys.stdout

class Region:
    '''A region across the alignent that will be evaluated'''

    def __init__(self, begin, end):
        self.begin = begin
        self.end = end
        self.dir = tempfile.TemporaryDirectory()
        self.input = self.dir.name + '/input'
        self.output = self.dir.name + '/output'

def prepare_alignment(region, sub):
    '''Process a region - create fasta file as input for probcons'''
    records = []

    for s in sub:
        seq = ''
           
        for i in range(region.begin, region.end):
            chr = sub[s][i]
            if i < region.end-args.word_size :
                sub[s][i] = ''
                seq += chr
            else :
                # subsequences that have gaps in the right anchor region are by definition
                # rare but need to be replaced by N's. This is important as later non-dashes
                # in the right anchor are removed to when the new msa is inserted.
                if chr == '-':
                    seq += 'n'
                else:
                    seq += chr

        record = SeqRecord.SeqRecord(
            Seq.Seq(seq.replace('-','')),
            id=' '.join([s, str(region.begin), str(region.end)]),
            description=''
        )
        records.append(record)
    FastaIO.FastaWriter(region.input, wrap=None).write_file(records)

def process_alignment(region, sub):
    '''After a region has been processed, incorporate the results'''
    for record in SeqIO.parse(region.output, 'fasta'):
        s = record.id
        res = list(record.seq)
        removed = 0

        for k in reversed(range(len(res))):
            if wp.search(res[k]):
                res[k] = ''
                removed += 1
                if removed == args.word_size:
                    break

        sub[s][region.begin] = ''.join(res).lower()

def run_job(fin, fout):
    '''run the job on one of our workers'''
    f = open(fout, 'w')
    subprocess.run(['probcons', '-c', '0', fin], stdout=f, check=True)

if __name__ == '__main__':

    if args.benchmark:
        start_time = time.time()

    # bracket our sequences with 22 ns on each side
    GGGS = 'n' * 22
    ANCHOR = 'G' * args.word_size

    # a dictionary of our sub reads
    sub = {}

    # the length of our alignmet
    alnlen = 0

    # read in the fasta file
    for record in SeqIO.parse(args.msa, 'fasta'):
        seq = GGGS + record.seq.upper() + GGGS
        sub[record.id] = list(seq)
        # build a thing based on the length of sequences
        alnlen = len(seq)

    # template for good positions
    good = ''

    for i in range(alnlen):
        base = Counter()
        for s in sub.values():
            base.update(s[i])
        best = base.most_common(1)[0]
        subcount = len(sub)

        # I changed the gap cutoff from >= 10% to >10% to simplify the logic
        if ( best[0] == '-' or best[1]/subcount < 0.9 ) :
            good += '-'
        else:
            good += 'G'

    # find regions of gaps, defined as gap to the first sequence of word_size G bases
    # we will keep a list of region coordinates
    regions = []

    # i will count our way across the alignment
    start = args.word_size

    # process all the gaps
    while (i := good.find('-', start)) > 0:

        # want an upstream anchor, default end is end of alignment
        begin = i - args.word_size
        end = alnlen

        # if there is an anchor, start after it
        if ( j := good.find(ANCHOR, i)) > 0:
            end = j + args.word_size

        # save the region
        regions.append(Region(begin,end))

        # update start
        start = end - args.word_size

    # regular expression for matching word
    wp = re.compile('\w')

    if args.scan:
        print(f'Found {len(regions)} regions with word-size {args.word_size}')
        sys.exit()

    sys.stderr.write(f'We have {len(regions)} regions to investigate')

    # farm out probcons to a worker pool, output is written to a file in a temp dir
    with mp.Pool(processes=args.max_cores) as pool:
        results = []
        for reg in regions:
            prepare_alignment(reg, sub)
            results.append(pool.apply_async(run_job, (reg.input, reg.output)))
        for res in results:
            res.get()

    for reg in regions:
        process_alignment(reg, sub)

    records = []
    for s in sub:
        record = SeqRecord.SeqRecord(
            Seq.Seq(''.join(sub[s]).replace('n','-')),
            id=s,
            description=''
        )
        records.append(record)

    FastaIO.FastaWriter(output, wrap=None).write_file(records)

    # finish
    if args.benchmark:
        end_time = time.time()
        with open(f'{args.base}-improve-msa.benchmark', 'a') as bf:
            bf.write(f'{start_time}\n')
            bf.write(f'{end_time}\n')
